### PR TITLE
[Train] Convert All doc tests and examples to use new Checkpoint API

### PR DESCRIPTION
**Pull request by woshiyyya**
## Why are these changes needed?

Convert most Ray Train/AIR/Tune examples and doc code to use new Checkpoint API.

We need to fix the rest in the following PR:
- `doc/source/train/examples/lightning/lightning_cola_advanced.ipynb`: LightningTrainer -> TorchTrainer
- `doc/source/ray-air/examples/dolly_lightning_fsdp_finetuning.ipynb`: Fix release test
- depends on #38866
  - `doc/source/ray-air/examples/lightgbm_example.ipynb` 
  - `doc/source/ray-air/examples/xgboost_example.ipynb`
  - `doc/source/tune/examples/tune_analyze_results.ipynb`

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/tune/doc_code/fault_tolerance.py [hunk 1]
```diff
@@ -1,22 +1,32 @@
 # flake8: noqa
 
 # __ft_initial_run_start__
+import pickle
+from tempdir import TemporaryDirectory
```
Comment (justinvyu) [inline doc/source/tune/doc_code/fault_tolerance.py:5]:
```suggestion
from tempfile import TemporaryDirectory
```

File: python/ray/train/examples/horovod/horovod_cifar_pbt_example.py [hunk 2]
```diff
@@ -111,14 +112,17 @@ def train_loop_per_worker(config):
...
+
+        with TemporaryDirectory() as tmpdir:
+            state_dict = dict(
                 model_state=net.state_dict(),
                 optimizer_state=optimizer.state_dict(),
                 epoch=epoch,
             )
-        )
-        train.report(dict(loss=running_loss / epoch_steps), checkpoint=checkpoint)
+            torch.save(state_dict, f"{tmpdir}/checkpoint.bin")
```
Comment (justinvyu) [inline python/ray/train/examples/horovod/horovod_cifar_pbt_example.py:123]: can we only use `os.path.join` for all examples?

Comment (justinvyu) [reply python/ray/train/examples/horovod/horovod_cifar_pbt_example.py:123]: Hopefully that's easy to do with find and replace 😅

File: doc/source/ray-air/examples/huggingface_text_classification.ipynb [hunk 3]
```diff
@@ -2152,9 +2152,9 @@
    },
    "outputs": [],
    "source": [
-    "from ray.train.huggingface import LegacyTransformersCheckpoint\n",
+    "from ray.train.huggingface import TransformersCheckpoint\n",
     "\n",
-    "checkpoint = LegacyTransformersCheckpoint.from_checkpoint(result.checkpoint)\n",
+    "checkpoint = TransformersCheckpoint.from_checkpoint(result.checkpoint)\n",
```
Comment (justinvyu) [inline doc/source/ray-air/examples/huggingface_text_classification.ipynb:2157]: This needs to be `TransformersCheckpoint(path=result.checkpoint.path, filesystem=result.checkpoint.filesystem)`

**Review by justinvyu — COMMENTED** [review]
Thanks, this is good progress!!

Btw, PBT is not landed yet so it's causing some errors. Also some issues I saw:

File: doc/source/ray-air/examples/huggingface_text_classification.ipynb (hunk 3 above)
Comment (woshiyyya) [inline doc/source/ray-air/examples/huggingface_text_classification.ipynb:2157]:
```suggestion
    "checkpoint = TransformersCheckpoint(path=result.checkpoint.path, filesystem=result.checkpoint.filesystem)\n",
```

File: python/ray/train/examples/horovod/horovod_cifar_pbt_example.py (hunk 2 above)
Comment (woshiyyya) [reply python/ray/train/examples/horovod/horovod_cifar_pbt_example.py:123]: ahh ok. Let me replace them,

**Review by krfricke — APPROVED** [review]

**Pull request closed without merging**